File: upwind.jl
type Upwind
    ndims::Int
    velocity::Array
    lengths::Array
    numCells::Array
    deltas::Array
    upDirection::Array
    ntot::Int
    dimProd::Array
    f::Array
end

function advect!(up::Upwind,deltaTime)
    oldF = deepcopy(up.f)
    for i = 0:up.ntot-1
        inds = getIndexSet(up,i)
        for j = 1:up.ndims
            oldIndex = inds[j]
            #periodic BCs
            inds[j] += up.upDirection[j] + up.numCells[j] 
            inds[j] %= up.numCells[j]
            upI = getFlatIndex(up,inds)
            up.f[i+1] -= (deltaTime * up.velocity[j] * up.upDirection[j]) * (oldF[upI+1]-oldF[i+1])/up.deltas[j]
            inds[j] = oldIndex
        end
    end 
end

function getFlatIndex(up::Upwind,inds)
    ans = Int(dot(up.dimProd,inds))
    return ans
end

function getIndexSet(up::Upwind,flatIndex)
    res = zeros(Int, up.ndims)
    for i = 0:up.ndims-1
        res[i+1] =  mod(Int(floor(flatIndex/up.dimProd[i+1])), up.numCells[i+1])
    end
    return res
end

function  saveVTK(up::Upwind)

end

function checksum(up::Upwind)
    return sum(up.f)
end

# Beginning of main program
ndims = 3

# Parse command line arguments
if length(ARGS) < 1
    println("Must specify number of cells in each direction.")
    println("Usage: ", basename(Base.source_path()), " numCells [numTimeSteps]")
    exit(1)
end
ncells = parse(Int, ARGS[1])

numTimeSteps = 100
if length(ARGS) > 1
    numTimeSteps = parse(Int, ARGS[2])
end

#Same resolution in each direction.
numCells = [ncells,ncells,ncells]
println(numCells)


#Initalizing vector velocity and lengths
velocity = ones(Float64,ndims)
lengths   = ones(Float64,ndims)


#computer dt
courant = 0.1
dt = Inf
dx = 0.00e0
for i = 1:ndims
    dx = lengths[i]/float(numCells[i])
    dt = min((courant*dx)/velocity[i],dt)
end


v = velocity
deltas = zeros(Float64,ndims)
upDirection = zeros(Integer,ndims)
ntot = 1
for i in 1:ndims
    upDirection[i] = -1
    if velocity[i] < 0.00
        upDirection[i] = +1
    end
    deltas[i] = lengths[i]/numCells[i]
    ntot *= numCells[i]
end

dimProd= ones(Integer, ndims)
for i = ndims-2:-1:0
    dimProd[i+1] = dimProd[i+2]*numCells[i+2]
end


f = zeros(ntot)
f[1] = 1.00e0
up = Upwind(ndims,velocity,lengths,numCells,deltas,upDirection,ntot,dimProd,f)


for i = 1:numTimeSteps
   advect!(up,dt) 
end 
println("Time evolution complete")


println("Check_sum= ",checksum(up))

File: upwind/cxx/upwind.cxx
#include <vector>
#include <functional>
#include <algorithm>
#include <numeric>
#include <string>
#include <fstream>
#include <limits>
#include <iostream>
#include <sstream>
#include <cstring>
#include "CmdLineArgParser.h"

#ifdef HAVE_OPENMP
#include <omp.h>
#endif

template <size_t NDIMS>
class Upwind {
public:
  // ctor
  Upwind(const std::vector<double>& velocity, 
         const std::vector<double>& lengths, 
         const std::vector<size_t>& numCells) {

    this->numCells = numCells;
    this->deltas.resize(NDIMS);
    this->upDirection.resize(NDIMS);
    this->v = velocity;
    this->lengths = lengths;
    this->ntot = 1;
    for (size_t j = 0; j < NDIMS; ++j) {
      this->upDirection[j] = -1;
      if (velocity[j] < 0.) this->upDirection[j] = +1;
      this->deltas[j] = lengths[j] / numCells[j];
      this->ntot *= numCells[j];
    }
    this->dimProd.resize(NDIMS);
    this->dimProd[NDIMS - 1] = 1;
    for (int i = (int) NDIMS - 2; i >= 0; --i) {
        // last index varies fastest
        this->dimProd[i] =  this->dimProd[i + 1] * this->numCells[i + 1];
      }
    this->f.resize(this->ntot, 0.0);
    this->oldF.resize(this->ntot, 0.0);
    // initialize lower corner to one
    this->f[0] = 1;
  }

  void advect(double deltaTime) {

    // copy
    this->oldF = this->f;
    const double* __restrict__ oldFPtr = &this->oldF[0];
    double* __restrict__ fPtr = &this->f[0];

#pragma omp parallel for
    for (int i = 0; i < (int) this->ntot; ++i) {

      std::vector<int> inds = this->getIndexSet(i);

      for (size_t j = 0; j < NDIMS; ++j) {

        int oldIndex = inds[j];
        const double coeff = deltaTime * this->v[j] * this->upDirection[j] / this->deltas[j];

        // periodic BCs
        inds[j] += this->upDirection[j] + this->numCells[j];
        inds[j] %= this->numCells[j];

        size_t upI = this->getFlatIndex(inds);

        fPtr[i] -= coeff * (oldFPtr[upI] - oldFPtr[i]);

        inds[j] = oldIndex;
      }
    }
  }

#include "saveVTK.h"


  double checksum() const {
    return std::accumulate(this->f.begin(), this->f.end(), 0.0, std::plus<double>());
  }

  void print() const {
    for (size_t i = 0; i < this->f.size(); ++i) {
      std::cout << i << " " << this->f[i] << '\n';
    }
  }

private:
  std::vector<double> v;
  std::vector<double> lengths;
  std::vector<double> deltas;
  std::vector<double> f;
  std::vector<double> oldF;
  std::vector<int> upDirection;
  std::vector<size_t> dimProd;
  std::vector<size_t> numCells;
  size_t ntot;

  inline
  std::vector<int> getIndexSet(size_t flatIndex) const {
    std::vector<int> res(NDIMS);
    for (size_t i = 0; i < NDIMS; ++i) {
      res[i] = flatIndex / this->dimProd[i] % this->numCells[i];
    }
    return res;
  }

  inline 
  size_t getFlatIndex(const std::vector<int>& inds) const {
    return std::inner_product(this->dimProd.begin(), this->dimProd.end(), inds.begin(), 0);
  }
};

///////////////////////////////////////////////////////////////////////

int main(int argc, char** argv) {

  const int ndims = 3;

  CmdLineArgParser args;
  args.setPurpose("Purpose: benchmark finite difference operations.");
  args.set("-numCells", 128, "Number of cells along each axis");
  args.set("-numSteps", 10, "Number of time steps");
  args.set("-vtk", false, "Write output to VTK file");

  bool success = args.parse(argc, argv);
  bool help = args.get<bool>("-h");

  if (success && !help) {

    int numTimeSteps = args.get<int>("-numSteps");
    bool doVtk = args.get<bool>("-vtk");

#pragma omp parallel
    {
      int numThreads = 1;
      int maxNumThreads = 1;
      int threadId = 0;
#ifdef HAVE_OPENMP
      numThreads = omp_get_num_threads();
      maxNumThreads = omp_get_max_threads();
      threadId = omp_get_thread_num();
      if (threadId == 0)
        std::cout << "Running with OpenMP enabled\n";
#endif
      if (threadId == 0)
        std::cout << "number of threads: " << 
           numThreads << " max number of threads: " << maxNumThreads << '\n';
    }

    // same resolution in each direction
    std::vector<size_t> numCells(ndims, args.get<int>("-numCells"));
    std::cout << "number of cells: ";
    for (size_t i = 0; i < numCells.size(); ++i) {
      std::cout << ' ' << numCells[i];
    }
    std::cout << '\n';
    std::cout << "number of time steps: " << numTimeSteps << '\n';

    std::vector<double> velocity(numCells.size(), 1.0);
    std::vector<double> lengths(numCells.size(), 1.0);

    // compute dt 
    double courant = 0.1;
    double dt = std::numeric_limits<double>::max();
    for (size_t j = 0; j < velocity.size(); ++j) {
      double dx = lengths[j]/numCells[j];
      double val = courant * dx / velocity[j];
      dt = (val < dt? val: dt);
    }

    Upwind<ndims> up(velocity, lengths, numCells);
    if (doVtk) {
      up.saveVTK("up0.vtk");
    }
    for (int i = 0; i < numTimeSteps; ++i) {
      up.advect(dt);
    }
    std::cout << "check sum: " << up.checksum() << '\n';
    if (doVtk) {
      up.saveVTK("up1.vtk");
    }
  }
  else {
    // error when parsing command line arguments
    if (!success) {
      std::cerr << "ERROR when parsing command line arguments\n";
    }
    args.help();
  }


  return 0;
}
